core/arch/arm/plat-imx/drivers/tzc380.c

```diff
@@ -11,17 +11,35 @@
 #include <imx.h>
 #include <mm/core_memprot.h>
 #include <mm/generic_ram_layout.h>
+#include <config.h>
+
+/*
+ * Used during tzc380 initialization, should be removed by dead code
+ * eliminiation when set to 0.
+ * This is needed to compile the code for i.MX6UL/L
+ * and i.MX8MQ.
+ */
+#ifndef TZASC2_BASE
+#define TZASC2_BASE			0
+#endif
 
 void imx_configure_tzasc(void)
 {
 
-	int i;
+	int i = 0;
+	int end = 1;
 	vaddr_t addr[2] = {0};
 
 	addr[0] = core_mmu_get_va(TZASC_BASE, MEM_AREA_IO_SEC);
-	addr[1] = core_mmu_get_va(TZASC2_BASE, MEM_AREA_IO_SEC);
 
-	for (i = 0; i < 2; i++) {
+	if (IS_ENABLED(CFG_MX6Q) || IS_ENABLED(CFG_MX6D)
+	    || IS_ENABLED(CFG_MX6DL)) {
+		assert(TZASC2_BASE != 0);
```

Comment:
@Emantor 
I don't see the interest of the panic during the boot. The code must not compile at least if the TZASC2 is mandatory on a device. This is related to the baseport if you need to configure something you must do it there is no choice.

Reply:
Sorry, I don't understand what you are trying to say, can you reformulate or give an example?

Reply:
With the assert, the boot is going to panic and boot will stop here.
Considering that the TZASC2_BASE must be defined in the platform using it, not defining it is a mistake of code implementation. Hence, compilation must not be possible in this case. 
If the compilation is fine whereas the code is wrong is not acceptable.

Reply:
`TZASC2_BASE` is now always defined as per the discussion with @sdininno, meaning it is defined to 0 for implementations which do not contain a `TZASC2_BASE`. The runtime assert is added for platforms which errornously choose this path, but rely on the fallback define to 0. All cases where `TZASC2_BASE` is defined to 0 should be removed by dead code elimination since the `if(IS_ENABLED...` before evaluates to `if(0)`. If they are not this is a bug, which should be fixed but currently can't be detected at compile time.

As said before, an implementation which ensures that all devices are detected correctly and at runtime is possible with device trees, but not in scope for this PR. This would also require the definition of device-tree bindings for the TZ380 and I'm not exactly sure where to keep them, since they are not specific to the kernel.

Reply:
1. Would `COMPILE_TIME_ASSERT()` help here?

2. About:
> device-tree bindings for the TZ380 and I'm not exactly sure where to keep them, since they are not specific to the kernel

I think the kernel is the authoritative place for all bindings at the moment, and there are tiny bits of secure stuff there already (see for instance [secure.txt](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/arm/secure.txt)). So it might be a good idea to get in touch with the kernel maintainers for DT bindings (Rob H., Mark R.. etc.), perhaps they would accept the bindings or suggest a better place.
That being said, I know there are discussions about the wider use of DT, see https://linaroconnectsandiego.sched.com/event/SufB/san19-115-system-device-trees.

Reply:
>     1. Would `COMPILE_TIME_ASSERT()` help here?

Sadly no, since the code explictly defines the `TZASC2_BASE` to 0. I.e. for `imx-imx8mqevk` the assert is done before the dead code is eliminated.
Also: the `COMPILE_TIME_ASSERT()` can probably be defined to `_Static_assert(...)` since OP-TEE is using `-std=gnu99`. But I don't know whether this is compatible with clang.

>     2. About:
> 
> 
> > device-tree bindings for the TZ380 and I'm not exactly sure where to keep them, since they are not specific to the kernel
> 
> I think the kernel is the authoritative place for all bindings at the moment, and there are tiny bits of secure stuff there already (see for instance [secure.txt](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/arm/secure.txt)). So it might be a good idea to get in touch with the kernel maintainers for DT bindings (Rob H., Mark R.. etc.), perhaps they would accept the bindings or suggest a better place.
> That being said, I know there are discussions about the wider use of DT, see https://linaroconnectsandiego.sched.com/event/SufB/san19-115-system-device-trees.

Thanks for the link, that looks interesting. Are the Linaro Connect events live streamed? Attending the Linaro Connect itself is unfortunately not possible.

Reply:
> Are the Linaro Connect events live streamed? 

AFAIK, only the keynote sessions are live streamed. Other presentations are recorded and uploaded to Youtube later.